python/lsst/ts/m2gui/controltab/tab_realtime_net_force_moment.py

```diff
@@ -0,0 +1,143 @@
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+__all__ = ["TabRealtimeNetForceMoment"]
+
+from PySide2.QtWidgets import QGridLayout
+from qasync import asyncSlot
+
+from ..display import FigureConstant
+from ..model import Model
+from .tab_default import TabDefault
+
+
+class TabRealtimeNetForceMoment(TabDefault):
+    """Table of the realtime net force and moment.
+
+    Parameters
+    ----------
+    title : `str`
+        Table's title.
+    model : `Model`
+        Model class.
+
+    Attributes
+    ----------
+    model : `Model`
+        Model class.
+    net_force_moment : `list`
+        Data of the net force and moment: [fx, fy, fz, mx, my, mz]. The unit of
+        force is Newton and the unit of moment is Newton * meter.
+    """
+
+    MIN_WIDTH = 150
+
+    def __init__(self, title: str, model: Model) -> None:
+        super().__init__(title, model)
+
+        self.net_force_moment = [0.0] * 6
+
+        # Realtime figures
+        self._figures = self._create_figures()
+
+        # Timer to update the realtime figures
+        self._timer = self.create_and_start_timer(self._callback_time_out)
+
+        self._set_widget_and_layout()
+
+    def _create_figures(self, num_realtime: int = 200) -> dict:
+        """Create the figures to show the net force and moment.
+
+        Parameters
+        ----------
+        num_realtime : `int`, optional
+            Number of the realtime data (>=0). (the default is 200)
+
+        Returns
+        -------
+        figures : `dict`
+            Figures of the net force and moment.
+        """
+
+        figures = dict()
+        axes = ["fx", "fy", "fz", "mx", "my", "mz"]
+        for axis in axes:
+            label_y = "Force (N)" if axis.startswith("f") else "Moment (N * m)"
+            figures[axis] = FigureConstant(
+                1,
+                num_realtime,
+                num_realtime,
+                "Data Point",
+                label_y,
+                axis,
+                legends=[],
+                num_lines=1,
+                is_realtime=True,
+            )
+
+        for figure in figures.values():
+            figure.axis_x.setLabelFormat("%d")
+
+        return figures
+
+    @asyncSlot()
+    async def _callback_time_out(self) -> None:
+        """Callback timeout function to update the realtime figures."""
+
+        for idx, figure in enumerate(self._figures.values()):
+            figure.append_data(self.net_force_moment[idx])
+
+        self.check_duration_and_restart_timer(self._timer)
+
+    def _set_widget_and_layout(self) -> None:
```

Comment:
This looks like a generic function. Is there any reason why it isn't in TabDefault? It is on ~5 classes, should be put into TabDefault:

```
Petrs-MacBook-Pro:ts_m2gui pkubanek$ grep -A10 -r "def _set_widget_and_layout" *
python/lsst/ts/m2gui/controltab/tab_actuator_control.py:    def _set_widget_and_layout(self) -> None:
python/lsst/ts/m2gui/controltab/tab_actuator_control.py-        """Set the widget and layout."""
python/lsst/ts/m2gui/controltab/tab_actuator_control.py-
python/lsst/ts/m2gui/controltab/tab_actuator_control.py-        widget = self.widget()
python/lsst/ts/m2gui/controltab/tab_actuator_control.py-        widget.setLayout(self._create_layout())
python/lsst/ts/m2gui/controltab/tab_actuator_control.py-
python/lsst/ts/m2gui/controltab/tab_actuator_control.py-        # Resize the dock to have a similar size of layout
python/lsst/ts/m2gui/controltab/tab_actuator_control.py-        self.resize(widget.sizeHint())
python/lsst/ts/m2gui/controltab/tab_actuator_control.py-
python/lsst/ts/m2gui/controltab/tab_actuator_control.py-        self.setWidget(self.set_widget_scrollable(widget, is_resizable=True))
python/lsst/ts/m2gui/controltab/tab_actuator_control.py-
--
python/lsst/ts/m2gui/controltab/tab_realtime_net_force_moment.py:    def _set_widget_and_layout(self) -> None:
python/lsst/ts/m2gui/controltab/tab_realtime_net_force_moment.py-        """Set the widget and layout."""
python/lsst/ts/m2gui/controltab/tab_realtime_net_force_moment.py-
python/lsst/ts/m2gui/controltab/tab_realtime_net_force_moment.py-        widget = self.widget()
python/lsst/ts/m2gui/controltab/tab_realtime_net_force_moment.py-        widget.setLayout(self._create_layout())
python/lsst/ts/m2gui/controltab/tab_realtime_net_force_moment.py-
python/lsst/ts/m2gui/controltab/tab_realtime_net_force_moment.py-        # Resize the dock to have a similar size of layout
python/lsst/ts/m2gui/controltab/tab_realtime_net_force_moment.py-        self.resize(widget.sizeHint())
python/lsst/ts/m2gui/controltab/tab_realtime_net_force_moment.py-
python/lsst/ts/m2gui/controltab/tab_realtime_net_force_moment.py-    def _create_layout(self) -> QGridLayout:
python/lsst/ts/m2gui/controltab/tab_realtime_net_force_moment.py-        """Create the layout.
--
python/lsst/ts/m2gui/controltab/tab_detailed_force.py:    def _set_widget_and_layout(self) -> None:
python/lsst/ts/m2gui/controltab/tab_detailed_force.py-        """Set the widget and layout."""
python/lsst/ts/m2gui/controltab/tab_detailed_force.py-
python/lsst/ts/m2gui/controltab/tab_detailed_force.py-        widget = self.widget()
python/lsst/ts/m2gui/controltab/tab_detailed_force.py-        widget.setLayout(self._create_layout())
python/lsst/ts/m2gui/controltab/tab_detailed_force.py-
python/lsst/ts/m2gui/controltab/tab_detailed_force.py-        self.setWidget(self.set_widget_scrollable(widget, is_resizable=True))
python/lsst/ts/m2gui/controltab/tab_detailed_force.py-
python/lsst/ts/m2gui/controltab/tab_detailed_force.py-    def _create_layout(self) -> QVBoxLayout:
python/lsst/ts/m2gui/controltab/tab_detailed_force.py-        """Create the layout.
python/lsst/ts/m2gui/controltab/tab_detailed_force.py-
--
python/lsst/ts/m2gui/controltab/tab_utility_view.py:    def _set_widget_and_layout(self) -> None:
python/lsst/ts/m2gui/controltab/tab_utility_view.py-        """Set the widget and layout."""
python/lsst/ts/m2gui/controltab/tab_utility_view.py-
python/lsst/ts/m2gui/controltab/tab_utility_view.py-        widget = self.widget()
python/lsst/ts/m2gui/controltab/tab_utility_view.py-        widget.setLayout(self._create_layout())
python/lsst/ts/m2gui/controltab/tab_utility_view.py-
python/lsst/ts/m2gui/controltab/tab_utility_view.py-        # Resize the dock to have a similar size of layout
python/lsst/ts/m2gui/controltab/tab_utility_view.py-        self.resize(widget.sizeHint())
python/lsst/ts/m2gui/controltab/tab_utility_view.py-
python/lsst/ts/m2gui/controltab/tab_utility_view.py-        self.setWidget(self.set_widget_scrollable(widget, is_resizable=True))
python/lsst/ts/m2gui/controltab/tab_utility_view.py-
--
python/lsst/ts/m2gui/controltab/tab_diagnostics.py:    def _set_widget_and_layout(self) -> None:
python/lsst/ts/m2gui/controltab/tab_diagnostics.py-        """Set the widget and layout."""
python/lsst/ts/m2gui/controltab/tab_diagnostics.py-
python/lsst/ts/m2gui/controltab/tab_diagnostics.py-        widget = self.widget()
python/lsst/ts/m2gui/controltab/tab_diagnostics.py-        widget.setLayout(self._create_layout())
python/lsst/ts/m2gui/controltab/tab_diagnostics.py-
python/lsst/ts/m2gui/controltab/tab_diagnostics.py-        # Resize the dock to have a similar size of layout
python/lsst/ts/m2gui/controltab/tab_diagnostics.py-        self.resize(widget.sizeHint())
python/lsst/ts/m2gui/controltab/tab_diagnostics.py-
python/lsst/ts/m2gui/controltab/tab_diagnostics.py-        self.setWidget(self.set_widget_scrollable(widget, is_resizable=True))
python/lsst/ts/m2gui/controltab/tab_diagnostics.py-
--
python/lsst/ts/m2gui/controltab/tab_config_view.py:    def _set_widget_and_layout(self) -> None:
python/lsst/ts/m2gui/controltab/tab_config_view.py-        """Set the widget and layout."""
python/lsst/ts/m2gui/controltab/tab_config_view.py-
python/lsst/ts/m2gui/controltab/tab_config_view.py-        widget = self.widget()
python/lsst/ts/m2gui/controltab/tab_config_view.py-        widget.setLayout(self._create_layout())
python/lsst/ts/m2gui/controltab/tab_config_view.py-
python/lsst/ts/m2gui/controltab/tab_config_view.py-        # Resize the dock to have a similar size of layout
python/lsst/ts/m2gui/controltab/tab_config_view.py-        self.resize(widget.sizeHint())
python/lsst/ts/m2gui/controltab/tab_config_view.py-
python/lsst/ts/m2gui/controltab/tab_config_view.py-    def _create_layout(self) -> QHBoxLayout:
python/lsst/ts/m2gui/controltab/tab_config_view.py-        """Create the layout.
--
python/lsst/ts/m2gui/controltab/tab_net_force_moment.py:    def _set_widget_and_layout(self) -> None:
python/lsst/ts/m2gui/controltab/tab_net_force_moment.py-        """Set the widget and layout."""
python/lsst/ts/m2gui/controltab/tab_net_force_moment.py-
python/lsst/ts/m2gui/controltab/tab_net_force_moment.py-        widget = self.widget()
python/lsst/ts/m2gui/controltab/tab_net_force_moment.py-        widget.setLayout(self._create_layout())
python/lsst/ts/m2gui/controltab/tab_net_force_moment.py-
python/lsst/ts/m2gui/controltab/tab_net_force_moment.py-        # Resize the dock to have a similar size of layout
python/lsst/ts/m2gui/controltab/tab_net_force_moment.py-        self.resize(widget.sizeHint())
python/lsst/ts/m2gui/controltab/tab_net_force_moment.py-
python/lsst/ts/m2gui/controltab/tab_net_force_moment.py-    def _create_layout(self) -> QVBoxLayout:
python/lsst/ts/m2gui/controltab/tab_net_force_moment.py-        """Create the layout.
```

Reply:
Fixed in the commit: [53db058](https://github.com/lsst-ts/ts_m2gui/pull/45/commits/53db05822cc77aec8c37891f1d6446ed80f12b1d). Thanks!